src/torchcodec/decoders/_core/VideoDecoder.cpp

```diff
@@ -916,6 +931,14 @@ void VideoDecoder::convertAVFrameToDecodedOutputOnCPU(
           "Invalid color conversion library: " +
           std::to_string(static_cast<int>(streamInfo.colorConversionLibrary)));
     }
+    if (!preAllocatedOutputTensor.has_value()) {
+      // We only convert to CHW if a pre-allocated tensor wasn't passed. When a
+      // pre-allocated tensor is passed, it's up to the caller (typically a
+      // batch API) to do the conversion. This is more efficient as it allows
+      // batch NHWC tensors to be permuted only once, instead of permuting HWC
+      // tensors N times.
+      output.frame = MaybeHWC2CHW(streamInfo.options, output.frame);
+    }
```

Comment:
I still think there's some smell to this. Whether the tensor was pre-allocated and whether it should be permuted should be orthogonal concepts.
I think it would make sense for all the low-level decoding capabilities (including this function `convertAVFrameToDecodedOutputOnCPU` ) to only ever accept and return HWC tensors.

And it should be up to the higher-level decoding entry-points (basically the moral equivalent of the public methods) to do the conversion. It's not trivial because `getFrameAtIndex` is *both* an entry point *and* a sub-function of other entry-points. Maybe that also means we should already let *all* entry-points do their own allocation and always pass pre-allocated tensors?

Reply:
Agreed on your reasoning and the principles.

One way to square the circle is to split the public facing part of `getFrameAtIndex` from the actual work being done. We would have a public member ('getFrameAtIndex'). function and a private member function ('getFrameAtIndexInternal`, or something like that).
  * `getFrameAtInedex` would call `getFrameAtIndexInternal` for the actual work, and after, it would do the conversion check and the actual conversion if needed.
  * `getFrameAtIndexInternal` would just assume HWC tensors and do the real work.

Then all of the internal calls to `getFrameAtIndex` would become calls to `getFrameAtIndexInternal`. I've implementing variants this pattern several times before. We would repeat this for any public entry points that also need to do work for other public entry points, as needed.

Reply:
I am fine with low-level functions only dealing with HWC. AFAICT, most (all?) low-level code deals with HWC because it has better performance.

Reply:
Yeah, that sounds good. I'll try to implement that in a follow-up PR

Reply:
> FAICT, most (all?) low-level code deals with HWC

That's not quite the case. In `main` , `convertAVFrameToDecodedOutputOnCPU` accepts both HWC and CHW - this is actually what this PR is fixing.

It may still return both HWC and CHW (instead of just HWC), and this is what I want to fix as a follow-up